File: tools/ais-runner/src/issues.ts
import type { RunnerWorkspaceIssue, RunnerWorkflowIssue } from './types.js';

export type StructuredIssue = {
  kind: string;
  severity: 'error' | 'warning' | 'info';
  node_id?: string;
  field_path: string;
  message: string;
  reference?: string;
  related?: { path?: string; node_id?: string; field_path?: string; reference?: string };
};

export function structuredFromWorkspaceIssues(issues: RunnerWorkspaceIssue[]): StructuredIssue[] {
  const out: StructuredIssue[] = [];
  for (const i of issues ?? []) {
    const severity = i.severity === 'warning' || i.severity === 'info' ? i.severity : 'error';
    out.push({
      kind: 'workspace_validation',
      severity,
      field_path: i.field_path ?? '(root)',
      message: i.message,
      reference: i.reference,
      related: i.related_path ? { path: i.related_path } : undefined,
    });
  }
  return out;
}

export function structuredFromWorkflowIssues(issues: RunnerWorkflowIssue[]): StructuredIssue[] {
  const out: StructuredIssue[] = [];
  for (const i of issues ?? []) {
    out.push({
      kind: 'workflow_validation',
      severity: 'error',
      node_id: i.nodeId,
      field_path: i.field ?? '(root)',
      message: i.message,
      reference: i.reference,
    });
  }
  return out;
}

export function zodPathToFieldPath(path: Array<string | number>): string {
  if (path.length === 0) return '(root)';
  let out = '';
  for (const seg of path) {
    if (typeof seg === 'number') {
      out += `[${seg}]`;
      continue;
    }
    if (!out) out = seg;
    else out += `.${seg}`;
  }
  return out;
}

File: rust/ais-rs/crates/ais-core/src/issues.rs
use crate::field_path::FieldPath;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum IssueSeverity {
    Error,
    Warning,
    Info,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct StructuredIssue {
    pub kind: String,
    pub severity: IssueSeverity,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub node_id: Option<String>,
    pub field_path: FieldPath,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reference: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub related: Option<Value>,
}

impl StructuredIssue {
    pub fn sort_stable(issues: &mut [Self]) {
        issues.sort_by(|left, right| {
            (
                left.severity,
                &left.kind,
                &left.field_path,
                &left.message,
                &left.node_id,
            )
                .cmp(&(
                    right.severity,
                    &right.kind,
                    &right.field_path,
                    &right.message,
                    &right.node_id,
                ))
        });
    }
}

#[cfg(test)]
#[path = "issues_test.rs"]
mod tests;
